### PR TITLE
Folio: optimized course reserve lookup

### PR DESCRIPTION
**Pull request by damien-git**
Currently the Folio driver is slow to find reserve information in findReserves(). This is mostly because for each reserve item, it looks for the item, the holdings and the instance in Folio. This is not necessary because we can use the returned copiedItem.instanceId or copiedItem.instanceHrid, depending on getBibIdType().
After this optimization, indexing the course reserves takes a third of the time it took before.

### CODE REVIEW

**Comment by demiankatz** [comment]
Thanks, @damien-git, this is great! The driver course reserves code was originally developed prior to the 1.0.0 release of mod-courses, and I'm not sure if the copiedItem elements existed at that time (though it's also entirely possible I just overlooked them)... either way, it looks like they've been there since the first official release, so this seems like a safe and obvious improvement.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php:1473]: If you wanted to make this even more concise, you could consider something like:

```
$idProperty = $this->getBibIdType() === 'hrid' ? 'instanceHrid' : 'instanceId';
$bibId = $item->copiedItem->$idProperty ?? null;
```

**Review by demiankatz — CHANGES_REQUESTED** [review]
Before merging this, it occurred to me that it could be simplified... but if you dislike my suggestion, feel free to say so, and I'm happy to merge as-is. :-)

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (damien-git) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php:1473]: Works for me, please edit as you like.

Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php:1473]: Thanks! It looks like I don't have write access to this branch, so I can't make changes directly. If you'd like me to, then please give me access. If not, I can just modify as part of the merge-to-dev process. Either way, I'm currently waiting to get some course data set up in my local test environment, and I'll give this a closer look once that's ready.

**Comment by demiankatz** [comment]
@damien-git, it seems like the remaining huge problem here is that retrieving the instructor data requires iterating through all the courses. My test environment has 6,000+ courses, and so it takes hours. I wonder if there are any plans to offer a better option here (or if there's already something we could do to improve performance). As it is, I still haven't been able to test this, because my last attempt got interrupted by a network outage, and I don't have enough time left today to start over again.

**Comment by demiankatz** [comment]
Thanks again, @damien-git. I was finally able to test this properly and it looks good, so I'm merging my simplified version as commit https://github.com/vufind-org/vufind/commit/82daaf82ef8568f7fd9e26c18839dab7888070c0.

**Comment by damien-git** [comment]
@demiankatz Sorry about the permission, I thought I already granted it earlier. You are right that there are still loops, and maybe it could be further optimized.

**Comment by demiankatz** [comment]
Thanks, @damien-git! Regarding the instructor loop, I eventually ended up adding caching logic to the driver so that when my tests got interrupted, I could resume from where I left off instead of having to start over. I didn't keep this code after I was finished testing, but maybe some degree of configurable caching might be useful as a future feature. I'm happy to share what I did if there's any interest, but I also feel that introducing a cache in this context could cause problems if the expiration time were not tuned correctly.

**Pull request closed without merging**
